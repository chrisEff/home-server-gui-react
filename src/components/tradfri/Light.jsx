'use strict'

const api = require('../../../config').api
api.url = `${api.protocol}://${api.host}:${api.port}`

const React = require('react')
const PropTypes = require('prop-types')

const LightTitle = require('./LightTitle')
const BrightnessSlider = require('./BrightnessSlider')
const ColorSwitcher = require('./ColorSwitcher')
const LightSwitch = require('./LightSwitch')

class Light extends React.Component {

	static propTypes = {
		bulb: PropTypes.object.isRequired,
	}

	constructor (props) {
		super(props)
		this.state = {bulb: props.bulb}
	}

	render () {
		const bulb = this.state.bulb
		const className = bulb.state ? 'light on' : 'light'
		return (
			<div className={className + ' ' + bulb.color}>
				<LightTitle title={bulb.name} onChange={this.updateName} />

				<LightSwitch state={bulb.state} onchange={this.updateState}/>
				<BrightnessSlider brightness={bulb.brightness} onchange={this.updateBrightness}/>
				<ColorSwitcher bulbType={bulb.bulbType} color={bulb.color} onchange={this.updateColor}/>
			</div>
		)
	}

	updateName = async (name) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/name/${name}?key=${api.key}`, {method: 'PUT'})

		bulb.name = name
		this.setState({bulb})
	}

	updateState = async (state) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/state/${state}?key=${api.key}`, {method: 'PUT'})

		bulb.state = state
		this.setState({bulb})
	}

	updateBrightness = async (brightness) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/brightness/${brightness}?key=${api.key}`, {method: 'PUT'})

		bulb.brightness = brightness
		this.setState({bulb})
	}

	updateColor = async (color) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/color/${color}?key=${api.key}`, {method: 'PUT'})

		bulb.color = color
		this.setState({bulb})
	}
}

module.exports = Light
