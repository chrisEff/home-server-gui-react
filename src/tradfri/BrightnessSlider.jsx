'use strict'

const React = require('react')
const PropTypes = require('prop-types')

const BrightnessSlider = ({brightness, onchange}) =>
	<input className='brightness-slider'
				 type='range' min='0' max='254' defaultValue={brightness}
				 onMouseUp={(e) => onchange(e.target.value)}/>

BrightnessSlider.propTypes = {
	brightness: PropTypes.number.isRequired,
	onchange: PropTypes.func,
}

module.exports = BrightnessSlider
