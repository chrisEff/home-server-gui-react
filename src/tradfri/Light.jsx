'use strict'

const config = require('../../config')

const React = require('react')
const BrightnessSlider = require('./BrightnessSlider')
const ColorSwitcher = require('./ColorSwitcher')
const LightSwitch = require('./LightSwitch')

class Light extends React.Component {
	constructor (props) {
		super(props)
		this.state = {bulb: props.bulb}
	}

	render () {
		const bulb = this.state.bulb;
		const className = bulb.state ? 'light on' : 'light'
		return (
			<div className={className + ' ' + bulb.color}>
				<span className={bulb.name.length > 15 ? 'name long' : 'name'}>{bulb.name}</span>

				<LightSwitch state={bulb.state} onchange={this.updateState}/>
				<BrightnessSlider brightness={bulb.brightness} onchange={this.updateBrightness}/>
				<ColorSwitcher bulbType={bulb.bulbType} color={bulb.color} onchange={this.updateColor}/>
			</div>
		)
	}

	updateState = async (state) => {
		const bulb = this.state.bulb
		await fetch(
			`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device/${bulb.id}/state/${state}?key=${config.api.key}`,
			{method: 'PUT'}
		)

		bulb.state = state
		this.setState({bulb})
	}

	updateBrightness = async (brightness) => {
		const bulb = this.state.bulb
		await fetch(
			`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device/${bulb.id}/brightness/${brightness}?key=${config.api.key}`,
			{method: 'PUT'}
		)

		bulb.brightness = brightness
		this.setState({bulb})
	}

	updateColor = async (color) => {
		const bulb = this.state.bulb
		await fetch(
			`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device/${bulb.id}/color/${color}?key=${config.api.key}`,
			{method: 'PUT'}
		)

		bulb.color = color
		this.setState({bulb})
	}
}

module.exports = Light
