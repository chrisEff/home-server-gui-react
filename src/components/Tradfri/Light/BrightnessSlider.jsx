'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {setDeviceBrightness} from '@/actions/tradfri'

const BrightnessSlider = ({deviceId, brightness, onChange}) =>
	<input
		style={styles.brightnessSlider}
		type='range'
		min='0'
		max='254'
		defaultValue={brightness}
		onMouseUp={(e) => onChange(deviceId, parseInt(e.target.value))}
	/>

BrightnessSlider.propTypes = {
	deviceId: PropTypes.number.isRequired,
	brightness: PropTypes.number.isRequired,
	onchange: PropTypes.func,
}

const styles = {
	brightnessSlider: {
		margin: '20px 0',
		width: '120px',
	},
}

const mapStateToProps = (state, ownProps) => ({
})
const mapDispatchToProps = dispatch => ({
	onChange: (id, brightness) => dispatch(setDeviceBrightness(id, brightness)),
})

export default connect(mapStateToProps, mapDispatchToProps)(BrightnessSlider)
export {BrightnessSlider}
