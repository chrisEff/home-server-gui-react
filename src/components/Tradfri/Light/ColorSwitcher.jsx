'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {setDeviceColor} from '@/actions/tradfri'

const ColorSwitcher = ({deviceId, bulbType, selected, onChange}) => {
	let colors = ['neutral']
	if (bulbType === 'rgb') {
		colors = ['red', 'green', 'blue', 'yellow', 'pink', 'purple', 'warm', 'neutral', 'cold']
	} else if (bulbType === 'white-spectrum') {
		colors = ['warm', 'neutral', 'cold']
	}

	return (
		<div style={styles.colorSwitcher}>
			{colors.map(color => (
				<div key={`${deviceId}-color-${color}`} style={styles.color(selected === color)} onClick={() => onChange(deviceId, color)}>
					<div style={{...styles.inner, ...styles.colors[color]}} />
				</div>
			))}
		</div>
	)
}

ColorSwitcher.propTypes = {
	deviceId: PropTypes.number.isRequired,
	bulbType: PropTypes.string.isRequired,
	selected: PropTypes.string,
	onchange: PropTypes.func,
}

const styles = {
	colorSwitcher: {
		fontSize: 0,
	},
	color: (active) => ({
		borderRadius: '20px',
		border: active ? '2px solid #eee' : '2px solid transparent',
		display: 'inline-block',
	}),
	inner: {
		borderRadius: '15px',
		margin: '3px',
		height: '29px',
		width: '29px',
	},
	colors: {
		warm:    {backgroundColor: '#efd275'},
		neutral: {backgroundColor: '#f1e0b5'},
		cold:    {backgroundColor: '#f5faf6'},
		red:     {backgroundColor: 'red'},
		green:   {backgroundColor: 'green'},
		blue:    {backgroundColor: 'blue'},
		yellow:  {backgroundColor: 'yellow'},
		pink:    {backgroundColor: 'deeppink'},
		purple:  {backgroundColor: 'mediumpurple'},
	},
}

const mapStateToProps = (state, ownProps) => ({
})
const mapDispatchToProps = dispatch => ({
	onChange: (id, color) => dispatch(setDeviceColor(id, color)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ColorSwitcher)
export {ColorSwitcher}
