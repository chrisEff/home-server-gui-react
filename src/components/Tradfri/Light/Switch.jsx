'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {StyleSheet, css} from 'aphrodite'

import {setDeviceState} from '@/actions/tradfri'

let Switch = ({deviceId, state, color, onChange}) =>
	<div
		className={css(state ? [styles.bulb, styles.on, styles[color]] : [styles.bulb])}
		onClick={() => onChange(deviceId, state ? 0 : 1)}>ðŸ’¡</div>


Switch.propTypes = {
	color: PropTypes.string,
	deviceId: PropTypes.number.isRequired,
	state: PropTypes.number.isRequired,
	onchange: PropTypes.func,
}

const styles = StyleSheet.create({
	bulb: {
		':before': {
			color: 'transparent',
			content: "'ðŸ’¡'",
			opacity: '0.3',
			position: 'absolute',
			fontSize: '80px',
			filter: 'none',
			height: '68px',
			overflow: 'hidden',
		},
		cursor: 'pointer',
		filter: 'grayscale(100%) blur(5px)',
		fontSize: 80,
		paddingTop: '15px',
		textAlign: 'center',
		userSelect: 'none',
	},
	on: {
		filter: 'none',
	},
	warm: {
		textShadow: '0 0 1px black, 0 0 50px #efd275',
		':before': {textShadow: '0 0 0 #efd275'},
	},
	neutral: {
		textShadow: '0 0 1px black, 0 0 50px #f1e0b5',
		':before': {textShadow: '0 0 0 #f1e0b5'},
	},
	cold: {
		textShadow: '0 0 1px black, 0 0 50px #f5faf6',
		':before': {textShadow: '0 0 0 #f5faf6'},
	},
	red: {
		textShadow: '0 0 1px black, 0 0 50px red',
		':before': {textShadow: '0 0 0 red'},
	},
	green: {
		textShadow: '0 0 1px black, 0 0 50px green',
		':before': {textShadow: '0 0 0 green'},
	},
	blue: {
		textShadow: '0 0 1px black, 0 0 50px blue',
		':before': {textShadow: '0 0 0 blue'},
	},
	yellow: {
		textShadow: '0 0 1px black, 0 0 50px yellow',
		':before': {textShadow: '0 0 0 yellow'},
	},
	pink: {
		textShadow: '0 0 1px black, 0 0 50px deeppink',
		':before': {textShadow: '0 0 0 deeppink'},
	},
	purple: {
		textShadow: '0 0 1px black, 0 0 50px mediumpurple',
		':before': {textShadow: '0 0 0 mediumpurple'},
	},
})

const mapStateToProps = (state, ownProps) => ({
})
const mapDispatchToProps = dispatch => ({
	onChange: (id, state) => dispatch(setDeviceState(id, state)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Switch)
export {Switch}
