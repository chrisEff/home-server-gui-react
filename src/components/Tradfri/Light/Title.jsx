'use strict'

import React, {useRef, useState} from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {StyleSheet, css} from 'aphrodite'

import {setDeviceName} from '@/actions/tradfri'

const Title = ({deviceId, title, onChange}) => {
	const [editMode, setEditMode] = useState(false)
	const titleInput = useRef(title)

	if (!editMode) {
		return (
			<span className={
				css(title.length > 15 ? [styles.title, styles.long] : [styles.title])
			} onClick={() => setEditMode(true)}>
				{title}
			</span>
		)
	} else {
		return (
			<div>
				<input type='text' ref={titleInput} style={{width: 110}} defaultValue={title} />
				<button onClick={() => { setEditMode(false); onChange(deviceId, titleInput.current.value) }}>OK</button>
			</div>
		)
	}
}

Title.propTypes = {
	deviceId: PropTypes.number.isRequired,
	title: PropTypes.string.isRequired,
	onChange: PropTypes.func,
}

const styles = StyleSheet.create({
	title: {
		':hover': {
			':after': {
				content: "' ✏️'",
				fontSize: '0.9em',
			},
		},
		display: 'block',
		height: '1.6em',
	},
	long: {
		fontSize: '0.8em',
		height: '2em',
	},
})

const mapStateToProps = (state, ownProps) => ({
})
const mapDispatchToProps = dispatch => ({
	onChange: (id, name) => dispatch(setDeviceName(id, name)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Title)
export {Title}
