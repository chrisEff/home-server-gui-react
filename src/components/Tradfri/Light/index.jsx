'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'

import Title from './Title'
import BrightnessSlider from './BrightnessSlider'
import ColorSwitcher from './ColorSwitcher'
import Switch from './Switch'

const Light = ({bulb}) => {
	const className = bulb.state ? 'light on' : 'light'
	return (
		<div style={styles.light} className={className + ' ' + bulb.color}>
			<Title deviceId={bulb.id} title={bulb.name} />
			<Switch deviceId={bulb.id} state={bulb.state} />
			{!!bulb.state && <BrightnessSlider deviceId={bulb.id} brightness={bulb.brightness} />}
			{!!bulb.state && <ColorSwitcher deviceId={bulb.id} bulbType={bulb.bulbType} selected={bulb.colors} />}
		</div>
	)
}

Light.propTypes = {
	id: PropTypes.number,
	bulb: PropTypes.object.isRequired,
}

const styles = {
	light: {
		height: '316px',
	},
}

const mapStateToProps = (state, ownProps) => ({
	bulb: state.tradfriDevices.find(device => device.id === ownProps.id),
})
const mapDispatchToProps = dispatch => ({
})

export default connect(mapStateToProps, mapDispatchToProps)(Light)
export {Light}
