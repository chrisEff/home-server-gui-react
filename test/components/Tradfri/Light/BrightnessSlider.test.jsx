'use strict'

import React from 'react'
import {BrightnessSlider} from '@/components/Tradfri/Light/BrightnessSlider'

import {shallow} from 'enzyme'

describe('BrightnessSlider', () => {
	it('should render correctly', () => {
		const wrapper = shallow(<BrightnessSlider deviceId={1} brightness={1} />)
		expect(wrapper).toMatchSnapshot()
	})
})
