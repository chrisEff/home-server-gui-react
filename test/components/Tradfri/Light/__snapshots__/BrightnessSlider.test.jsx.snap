// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrightnessSlider should render correctly 1`] = `
<input
  defaultValue={1}
  max="254"
  min="0"
  onMouseUp={[Function]}
  style={
    Object {
      "margin": "20px 0",
      "width": "120px",
    }
  }
  type="range"
/>
`;
